contracts/hooks/examples/FullRange.sol

```diff
@@ -0,0 +1,358 @@
+            afterDonate: false
+        });
+    }
+
+    function addLiquidity(AddLiquidityParams calldata params)
+        external
+        ensure(params.deadline)
+        returns (uint128 liquidity)
+    {
+        PoolKey memory key = PoolKey({
+            currency0: Currency.wrap(params.token0),
+            currency1: Currency.wrap(params.token1),
+            fee: params.fee,
+            tickSpacing: 60,
+            hooks: IHooks(address(this))
+        });
+
+        PoolId poolId = key.toId();
+
+        (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(poolId);
+
+        if (sqrtPriceX96 == 0) revert PoolNotInitialized();
+
+        PoolInfo storage pool = poolInfo[poolId];
+
+        uint128 poolLiquidity = poolManager.getLiquidity(poolId);
+
+        liquidity = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtPriceX96,
+            TickMath.getSqrtRatioAtTick(MIN_TICK),
+            TickMath.getSqrtRatioAtTick(MAX_TICK),
+            params.amount0Desired,
+            params.amount1Desired
+        );
+
+        if (poolLiquidity == 0 && liquidity <= MINIMUM_LIQUIDITY) {
+            revert LiquidityDoesntMeetMinimum();
+        }
+        BalanceDelta addedDelta = modifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: liquidity.toInt256()
+            })
+        );
+
+        if (poolLiquidity == 0) {
+            // permanently lock the first MINIMUM_LIQUIDITY tokens
+            liquidity -= MINIMUM_LIQUIDITY;
+            UniswapV4ERC20(pool.liquidityToken).mint(address(0), MINIMUM_LIQUIDITY);
+        }
+
+        UniswapV4ERC20(pool.liquidityToken).mint(params.to, liquidity);
+
+        require(
+            uint128(addedDelta.amount0()) >= params.amount0Min && uint128(addedDelta.amount1()) >= params.amount1Min,
+            "Price slippage check"
+        );
+    }
+
+    function removeLiquidity(RemoveLiquidityParams calldata params)
+        public
+        virtual
+        ensure(params.deadline)
+        returns (BalanceDelta delta)
+    {
+        PoolKey memory key = PoolKey({
+            currency0: Currency.wrap(params.token0),
+            currency1: Currency.wrap(params.token1),
+            fee: params.fee,
+            tickSpacing: 60,
+            hooks: IHooks(address(this))
+        });
+
+        PoolId poolId = key.toId();
+
+        (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(poolId);
+
+        if (sqrtPriceX96 == 0) revert PoolNotInitialized();
+
+        UniswapV4ERC20 erc20 = UniswapV4ERC20(poolInfo[poolId].liquidityToken);
+
+        delta = modifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: -(params.liquidity.toInt256())
+            })
+        );
+
+        erc20.burn(msg.sender, params.liquidity);
```

Comment:
this is an interesting line. There's the mantra "checks, effects, interactions" to avoid reentrancy bugs by putting code in that order. burning the token I would say is an "effect" (even tho it's kind of an interaction with the token) bc we're changing the balance the LP is owed by this contract. I'm trying to think if there's a way they could reenter and remove liquidity again having the same token balance.

But I don't think there's a good place where the user can reenter, and we'd always hit this line multiple times and require all the liquidity so I don't think reentrancy matters. But it stood out since we're paying them what they're owed before we nil out the balance.

Reply:
either way tho, having this on line 176 might be preferable and it will cause an earlier revert if they try and burn more than they own